### DISCUSSION TITLE
Nested Relations in subdocuments

### DISCUSSION THREAD

**Discussion by cdelacruzpinto** [BeanieODM/beanie#723] [Feature Request]
Currently, the following case works with beanie. I can retrieve a list of todos using the get(fetch_links=True) method.
```
class Todo(Document):
  id: PydanticObjectId
  description: str

class TodoList(Document):
  owner: str
  todo_items = list[Link[Todo]]
```
So far, so good. But let's say I would have this instead:

```
class TodoList(Document):
   owner: str
   todos: List[TodoItem]

class TodoItem(BaseModel, ??no SubDocument class available):
  order: int
  todo: Link[Todo]
```
A document of the TodoList  would look in the db lke this:
```
{
  owner: "Carlos"
  todos:[
    {"order":1, "todo":$oid(000000001)}, {"order":2, "todo":$oid(0000000003}
  ]
}
```
I would have no way to retrieve those links without having to iterate the list of todo_items and call fetch for each item, which is super slow.

So my feature reqest would be:

"Support automatically fetching related objects for Links in nested SubDocuments, or document an efficient alternative" :-)

**Comment by mrrobinhood5** [comment]
`fetch_links=True` works if you make `TodoItem` a `Document` instead of a BaseModel. It does fetch Links up to about 4 or 5 levels deep

**Reply by akloster** [reply]
First, in my opinion, Todo should be embedded into TodoList. You can then still index the todos and make queries on the todos irrespective of the lists and so on. The only thing not possible is to assign one Todo to multiple lists.

It would be truly awkward to have the TodoItem as Document as well. Then you'd have three collections. And a query on the TodoList collection would need two $lookup stages. Ideally TodoItem would be embedded in TodoList, at the very least. I don't see how to do that in Beanie right now. In order for "Link" to work properly, it must be used in a Document, not a Pydantic model.

You could achieve something similar by embedding the TodoItem as a Pydantic model, and just referencing the Todo's id. Then you would use a custom aggregation to look those up. Many of Beanie's "relation" features won't work that way though.

Alternatively use two properties in the TodoList, one with a list of links to the Todos, and a list of numbers for the sort orders.

Also: Did you notice that list properties are already "ordered"? Putting a sort_order column into a table is common in SQL databases, since otherwise you can't really let the user customize the sort order. But because the list of links is already ordered and you can shuffle around that list all you want, you don't need that property at all.

Otherwise the best way may still be to have three Documents/collections. It's a very SQL-ish way of doing things though.

As far as "feature requests" from this little exercise: Maybe make it possible to fetch_links across pydantic models. Maybe create custom Relation classes, which can generate the appropriate $lookup stage, but may not need a "ref" (combination of collection and id), and may include properties not needed for the lookup (like [sort_order, id]). But both seems complicated to implement and not THAT useful.

**Comment by cdelacruzpinto** [comment]
First of all, thanks to both for taking the time answering my feature request with such a detailed answer. Also to @roman-right  for the awesome ODM :-).

The reason I wanted this 'weird' structure is just because I was trying to write a small, read-only service to retrieve data from an already existing collection. It has a similar structure to the one I have described/oversimplified. 

So after realizing that what I wanted is already partially possible, I guess that the feature request would be "allow to specify links that are only an ObjectID, letting the developer specify the target collection". This would allow more conveniently reading with Beanie data saved by MongoEngine, which creates links just using an ObjectID. But I can already see that this clashes with Beanie's design. I guess my only chance is using an aggregation pipeline. 

Thanks again to both, and I wish you a great weekend.
